File: book/1.1/solution.scala
val theList: List[Int] = List(1, 2, 3, 4, 5)
val answer: List[Int] = List(120, 60, 40, 30, 24)

// Simple reduction and map
def getProductListDivision: List[Int] => List[Int] = list => {
  val product = list.product // list.reduceLeft(_ * _) list.foldLeft(1)(_ * _)
  list map (product / _)
}

// Tail recursive function that multiplies prefix elements to suffix elements
// and keeps track with accumulator
@scala.annotation.tailrec
def getProductListRec(rest: List[Int], prefix: List[Int], acc: List[Int]): List[Int] =
  rest match {
    case Nil => acc
    case x :: xs => {
      val newPrefix = prefix :+ x
      val newAccum = acc :+ (prefix.product * xs.product)
      getProductListRec(xs, newPrefix, newAccum)
    }
  }

println( getProductListDivision(theList) )
println( getProductListRec(theList, Nil, Nil) )

File: book/1.1/solution.swift
let theList: [Int] = [1, 2, 3, 4, 5]
let answer: [Int] = [120, 60, 40, 30, 24]

// Just a helper to make this reduction more terse inline below
func getProduct(list: [Int]) -> Int {
  return list.reduce(1, {x, y in x * y})
}

// Simple reduction and map
func getProductListDivision(list: [Int]) -> [Int] {
  let product = getProduct(list: list)
  return list.map { product / $0 }
}

// Tail recursive function that multiplies prefix elements to suffix elements
// and keeps track with accumulator
func getProductListRec(rest: [Int], prefix: [Int], acc: [Int]) -> [Int] {
  if let x = rest.first {
    let xs = Array(rest.dropFirst(1)) // cast ArraySlice<Int> into [Int]
    let newPrefix = prefix + [x]
    let newAccum = acc + [(getProduct(list: prefix) * getProduct(list: xs))]
    return getProductListRec(rest: xs, prefix: newPrefix, acc: newAccum)
  }
  else {
    return acc
  }
}

print(getProductListDivision(list: theList))
print(getProductListRec(rest: theList, prefix: [], acc: []))